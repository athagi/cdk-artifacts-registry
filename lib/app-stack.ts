import * as cdk from '@aws-cdk/core';
import ec2 = require('@aws-cdk/aws-ec2');
// import * as ecr from '@aws-cdk/aws-ecr';

// class AppStack extends cdk.Stack {
export class AppStack extends cdk.Stack {
  public readonly vpc: ec2.Vpc;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // this.vpc = new ec2.Vpc(this, 'Vpc', { maxAzs: 1 });
  }
}
