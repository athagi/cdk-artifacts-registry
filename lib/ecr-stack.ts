import * as ecr from "@aws-cdk/aws-ecr";
import * as cdk from "@aws-cdk/core";

// ecr stack
export interface RepostiroiesStacksProps extends cdk.StackProps {
  repoNames: string[];
  lifecycleRule: ecr.LifecycleRule;
  prefix?: string;
}

export class RepositoriesStack extends cdk.Stack {
  public readonly repository: ecr.Repository;
  static LIFECYCLERULE: ecr.LifecycleRule = {
    maxImageCount: 10,
    rulePriority: 1,
    tagStatus: ecr.TagStatus.ANY,
  };

  constructor(
    scope: cdk.Construct,
    id: string,
    props: RepostiroiesStacksProps
  ) {
    super(scope, id, props);

    const repositories: string[] = props.repoNames;
    let prefix = "";
    if (
      props.prefix &&
      props.prefix.length > 0 &&
      !props.prefix.match(/(master|prod|production|main)/)
    ) {
      prefix = `${props.prefix}-`;
    }
    repositories.forEach((repo) => {
      this.createRepository(`${prefix}${repo}`, props.lifecycleRule);
    });
  }

  createRepository(
    repoName: string,
    lifecycleRule: ecr.LifecycleRule
  ): ecr.Repository {
    const repository = new ecr.Repository(this, repoName, {
      imageScanOnPush: true,
      repositoryName: repoName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    repository.addLifecycleRule(lifecycleRule);
    return repository;
  }
}
