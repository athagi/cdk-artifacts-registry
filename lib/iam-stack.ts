import * as iam from "@aws-cdk/aws-iam";
import * as cdk from "@aws-cdk/core";
export interface IamUserStacksProps extends cdk.StackProps {
  userNames: string[];
  strictedCidrs: string[];
  groupName: string;
  prefix?: string;
}
interface policyDocumentType {
  Version: string;
  Statement: policyDocumentStateType[];
}
interface policyDocumentStateType {
  Sid: string;
  Effect: string;
  Action: string[];
  Resource: string;
  Condition: ipAddressType;
}
interface ipAddressType {
  IpAddress: sourceIpType;
}
interface sourceIpType {
  "aws:SourceIp": string[];
}
export class IamUserStack extends cdk.Stack {
  private CIDR_REGEX = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/(3[0-2]|[12]?[0-9])$/;
  constructor(scope: cdk.Construct, id: string, props: IamUserStacksProps) {
    super(scope, id, props);
    props.strictedCidrs.forEach((ip) => {
      if (!this.CIDR_REGEX.test(ip)) {
        throw new Error("invalid ips");
      }
    });
    let prefix = "";
    if (
      props.prefix &&
      props.prefix.length > 0 &&
      !props.prefix.match(/(master|prod|production|main)/)
    ) {
      prefix = `${props.prefix}-`;
    }
    const group: iam.Group = new iam.Group(
      this,
      `${prefix}${props.groupName}`,
      {
        groupName: `${prefix}${props.groupName}`,
      }
    );
    const baseGroupInlinePolicy: iam.Policy = new iam.Policy(
      this,
      group + "InlinePolicy",
      {
        document: iam.PolicyDocument.fromJson(
          this.generateDefaultPolicyJson(props.strictedCidrs)
        ),
      }
    );
    group.attachInlinePolicy(baseGroupInlinePolicy);
    const users = props.userNames;
    users.forEach((u) => {
      const userName = `${prefix}${u}`;
      const user: iam.User = this.createIamUser(userName, group);
      const inlinePolicy: iam.Policy = new iam.Policy(
        this,
        userName + "-policy",
        {
          document: iam.PolicyDocument.fromJson(
            this.generateInlinePolicyJson(props.strictedCidrs, userName)
          ),
        }
      );
      user.attachInlinePolicy(inlinePolicy);
    });
  }
  createIamUser(userName: string, group: iam.Group): iam.User {
    const user: iam.User = new iam.User(this, userName, {
      userName: userName,
    });
    user.addToGroup(group);
    return user;
  }
  generateInlinePolicyJson(
    ips: string[],
    username: string
  ): policyDocumentType {
    const userPolicyDocument = {
      Version: "2008-10-17",
      Statement: [
        {
          Sid: "AllowPush",
          Effect: "Allow",
          Action: [
            "ecr:CompleteLayerUpload",
            "ecr:InitiateLayerUpload",
            "ecr:PutImage",
            "ecr:UploadLayerPart",
            "ecr-public:CompleteLayerUpload",
            "ecr-public:InitiateLayerUpload",
            "ecr-public:PutImage",
            "ecr-public:UploadLayerPart",
          ],
          Resource: `arn:aws:ecr:*:*:repository/${username}/*`,
          Condition: {
            IpAddress: {
              "aws:SourceIp": ips,
            },
          },
        },
      ],
    };
    return userPolicyDocument;
  }
  generateDefaultPolicyJson(ips: string[]): policyDocumentType {
    const policyDocument = {
      Version: "2012-10-17",
      Statement: [
        {
          Sid: "VisualEditor0",
          Effect: "Allow",
          Action: [
            "ecr:DescribeImageScanFindings",
            "ecr:GetLifecyclePolicyPreview",
            "ecr:GetDownloadUrlForLayer",
            "ecr:BatchGetImage",
            "ecr:DescribeImages",
            "ecr:GetAuthorizationToken",
            "ecr:DescribeRepositories",
            "ecr:ListTagsForResource",
            "ecr:ListImages",
            "ecr:BatchCheckLayerAvailability",
            "ecr:GetRepositoryPolicy",
            "ecr:GetLifecyclePolicy",
            "ecr-public:GetAuthorizationToken",
            "sts:GetServiceBearerToken",
            "ecr-public:BatchCheckLayerAvailability",
            "ecr-public:GetRepositoryPolicy",
            "ecr-public:DescribeRepositories",
            "ecr-public:DescribeRegistries",
            "ecr-public:DescribeImages",
            "ecr-public:DescribeImageTags",
            "ecr-public:GetRepositoryCatalogData",
            "ecr-public:GetRegistryCatalogData",
          ],
          Resource: "*",
          Condition: {
            IpAddress: {
              "aws:SourceIp": ips,
            },
          },
        },
      ],
    };
    return policyDocument;
  }
}
