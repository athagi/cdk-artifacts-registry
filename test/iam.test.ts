import { expect as expectCDK, matchTemplate, MatchStyle, haveResource, countResources, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Iam from '../lib/iam-stack';

describe('snapshot test', () => {
  it("iam is created", () => {
    const app = new cdk.App();
    // WHEN
    const userNames = ["user1", "user2"]
    const groupName = "test-group";
    const stack = new Iam.IamUserStack(app, 'MyTestStack',  {
      userNames: userNames, 
      strictedIps: ["0.0.0.0/0"],
      groupName: groupName,});
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  })
})

describe('fine grained tests', () => {
  it('User Created', () => {
    const app = new cdk.App();
    // WHEN
    const userNames = ["user1", "user2"]
    const groupName = "test-group";
    const stack = new Iam.IamUserStack(app, 'MyTestStack',  {
      userNames: userNames, 
      strictedIps: ["0.0.0.0/0"],
      groupName: groupName,});
    // THEN
    userNames.forEach(user => {
      expectCDK(stack).to(haveResource("AWS::IAM::User", {
        UserName: user,
      }));
    });
  });

  it('Group Created', () => {
    const app = new cdk.App();
    const userNames = ["user1"];
    const groupName = "test-group";
    const stack = new Iam.IamUserStack(app, 'MyTestStack',  {
      userNames: userNames, 
      strictedIps: ["0.0.0.0/0"],
      groupName: groupName,});
    expectCDK(stack).to(haveResource("AWS::IAM::Group", {
      "GroupName": groupName,
    }));
  });



})

