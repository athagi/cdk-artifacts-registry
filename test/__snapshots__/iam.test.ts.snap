// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test iam is created 1`] = `
Object {
  "Resources": Object {
    "MyTestStacktesttestgroupInlinePolicy21AA5C8E": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testtestgroup35B39A5B",
          },
        ],
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:DescribeImageScanFindings",
                "ecr:GetLifecyclePolicyPreview",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "ecr:DescribeImages",
                "ecr:GetAuthorizationToken",
                "ecr:DescribeRepositories",
                "ecr:ListTagsForResource",
                "ecr:ListImages",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetRepositoryPolicy",
                "ecr:GetLifecyclePolicy",
                "ecr-public:GetAuthorizationToken",
                "sts:GetServiceBearerToken",
                "ecr-public:BatchCheckLayerAvailability",
                "ecr-public:GetRepositoryPolicy",
                "ecr-public:DescribeRepositories",
                "ecr-public:DescribeRegistries",
                "ecr-public:DescribeImages",
                "ecr-public:DescribeImageTags",
                "ecr-public:GetRepositoryCatalogData",
                "ecr-public:GetRegistryCatalogData",
              ],
              "Condition": Object {
                "IpAddress": Object {
                  "aws:SourceIp": Array [
                    "0.0.0.0/0",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "VisualEditor0",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestStacktesttestgroupInlinePolicy21AA5C8E",
      },
      "Type": "AWS::IAM::Policy",
    },
    "testtestgroup35B39A5B": Object {
      "Properties": Object {
        "GroupName": "test-test-group",
      },
      "Type": "AWS::IAM::Group",
    },
    "testuser19D7501B4": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testtestgroup35B39A5B",
          },
        ],
        "UserName": "test-user1",
      },
      "Type": "AWS::IAM::User",
    },
    "testuser1policyAF1C3FE3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:CompleteLayerUpload",
                "ecr:InitiateLayerUpload",
                "ecr:PutImage",
                "ecr:UploadLayerPart",
                "ecr-public:CompleteLayerUpload",
                "ecr-public:InitiateLayerUpload",
                "ecr-public:PutImage",
                "ecr-public:UploadLayerPart",
              ],
              "Condition": Object {
                "IpAddress": Object {
                  "aws:SourceIp": Array [
                    "0.0.0.0/0",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": "arn:aws:ecr:*:*:repository/test-user1/*",
              "Sid": "AllowPush",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testuser1policyAF1C3FE3",
        "Users": Array [
          Object {
            "Ref": "testuser19D7501B4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testuser23C5E23F4": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testtestgroup35B39A5B",
          },
        ],
        "UserName": "test-user2",
      },
      "Type": "AWS::IAM::User",
    },
    "testuser2policy70FDC9F5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:CompleteLayerUpload",
                "ecr:InitiateLayerUpload",
                "ecr:PutImage",
                "ecr:UploadLayerPart",
                "ecr-public:CompleteLayerUpload",
                "ecr-public:InitiateLayerUpload",
                "ecr-public:PutImage",
                "ecr-public:UploadLayerPart",
              ],
              "Condition": Object {
                "IpAddress": Object {
                  "aws:SourceIp": Array [
                    "0.0.0.0/0",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": "arn:aws:ecr:*:*:repository/test-user2/*",
              "Sid": "AllowPush",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testuser2policy70FDC9F5",
        "Users": Array [
          Object {
            "Ref": "testuser23C5E23F4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
